base_layer/core/src/consensus/consensus_encoding/hashing.rs

```diff
@@ -20,87 +20,93 @@
 //  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 //  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-use std::io::Write;
+use std::{io, io::Write, marker::PhantomData};
 
-use digest::{consts::U32, Digest, FixedOutput, Update};
+use digest::{consts::U32, Digest};
 use tari_common_types::types::HashDigest;
+use tari_crypto::hashing::{DomainSeparatedHasher, DomainSeparation};
 
 use crate::consensus::ConsensusEncoding;
 
+/// Domain separated consensus encoding hasher.
+pub struct DomainSeparatedConsensusHasher<M: DomainSeparation>(PhantomData<M>);
+
+impl<M: DomainSeparation> DomainSeparatedConsensusHasher<M> {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(label: &'static str) -> ConsensusHasher<DomainSeparatedHasher<HashDigest, M>> {
+        ConsensusHasher::new(DomainSeparatedHasher::new_with_label(label))
+    }
+}
+
 #[derive(Clone)]
-pub struct ConsensusHashWriter<H> {
-    digest: H,
+pub struct ConsensusHasher<D> {
+    writer: WriteHashWrapper<D>,
 }
 
-impl<H: Digest> ConsensusHashWriter<H> {
-    pub fn new(digest: H) -> Self {
-        Self { digest }
+impl<D: Digest> ConsensusHasher<D> {
+    pub fn new(digest: D) -> Self {
+        Self {
+            writer: WriteHashWrapper(digest),
+        }
     }
 }
 
-impl<H> ConsensusHashWriter<H>
-where H: FixedOutput<OutputSize = U32> + Update
+impl<D> ConsensusHasher<D>
+where D: Digest<OutputSize = U32>
 {
     pub fn finalize(self) -> [u8; 32] {
-        self.digest.finalize_fixed().into()
+        self.writer.0.finalize().into()
     }
 
-    pub fn update_consensus_encode<T: ConsensusEncoding>(&mut self, data: &T) {
+    pub fn update_consensus_encode<T: ConsensusEncoding + ?Sized>(&mut self, data: &T) {
         // UNWRAP: ConsensusEncode MUST only error if the writer errors, HashWriter::write is infallible
-        data.consensus_encode(self)
+        data.consensus_encode(&mut self.writer)
             .expect("Incorrect implementation of ConsensusEncoding encountered. Implementations MUST be infallible.");
     }
 
     pub fn chain<T: ConsensusEncoding>(mut self, data: &T) -> Self {
         self.update_consensus_encode(data);
         self
     }
+}
 
-    pub fn into_digest(self) -> H {
-        self.digest
+impl Default for ConsensusHasher<HashDigest> {
+    fn default() -> Self {
+        ConsensusHasher::new(HashDigest::new())
     }
 }
 
-impl<H: Update> Write for ConsensusHashWriter<H> {
-    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
-        self.digest.update(buf);
+/// This private struct wraps a Digest and implements the Write trait to satisfy the consensus encoding trait..
+#[derive(Clone)]
```

Comment:
Couldn't we provide a blanket `Write` impl on the underlying `DomainHasher`?

Reply:
In this case we can't because the API for `Write` is meant to allow "chunked" writes and the `DomainSeparatedHasher` does not allow this.
```
c = a||b
H.chain(a).chain(b) != H.chain(a||b)
```

 This is the main reason why I chose to make the write interface internal to the hasher, because we know that ConseususEncoding writes each element in its entirety.